src/libPMacc/include/cuSTL/algorithm/mpi/Gather.tpp

```diff
@@ -146,22 +173,66 @@ int Gather<dim>::rank() const
 }
 
 template<int dim>
-template<typename Type, int memDim, class T_Alloc, class T_Copy, class T_Assign, class T_Alloc2, class T_Copy2, class T_Assign2>
-void Gather<dim>::CopyToDest::operator()(const Gather<dim>& gather,
+template<typename Type, int memDim, class T_Alloc, class T_Copy, class T_Assign>
+void Gather<dim>::CopyToDest::operator()(
+                        const Gather<dim>& gather,
                         container::CartBuffer<Type, memDim, T_Alloc, T_Copy, T_Assign>& dest,
                         std::vector<Type>& tmpDest,
-                        container::CartBuffer<Type, memDim, T_Alloc2, T_Copy2, T_Assign2>& source, int dir) const
+                        int dir,
+                        const std::vector<math::Size_t<memDim> >& srcSizes,
+                        const std::vector<size_t>& srcOffsets1D) const
 {
     using namespace math;
 
-    for(int i = 0; i < (int)gather.positions.size(); i++)
+    int numRanks = (int)gather.positions.size();
+
+    // calculate offsets per axis in destination buffer
+    std::vector<size_t> offsets_per_axis[memDim];
+    std::vector<size_t> sizes_per_axis[memDim];
+
+    // sizes per axis
+    for(int i = 0; i < numRanks; i++)
     {
         Int<dim> pos = gather.positions[i];
         Int<memDim> posInMem = GatherHelper::posInMem<dim, memDim>()(pos, dir);
+        for(int axis = 0; axis < memDim; axis++)
+        {
+            size_t pos_on_axis = (size_t)posInMem[axis];
+            if(pos_on_axis >= sizes_per_axis[axis].size())
+                sizes_per_axis[axis].resize(pos_on_axis + 1);
+            sizes_per_axis[axis][pos_on_axis] = srcSizes[i][axis];
+        }
+    }
+
+    // offsets per axis
+    for(int axis = 0; axis < memDim; axis++)
+    {
+        offsets_per_axis[axis].resize(sizes_per_axis[axis].size());
+        std::vector<size_t> partialSum(offsets_per_axis[axis].size());
+        std::partial_sum(sizes_per_axis[axis].begin(), sizes_per_axis[axis].end(), partialSum.begin());
+        offsets_per_axis[axis][0] = 0;
+        std::copy(partialSum.begin(), partialSum.end()-1, offsets_per_axis[axis].begin()+1);
+    }
+
+    // copy from 1-dim mpi buffer to ndim destination buffer
```

Comment:
I can't understand this comment. 
Copy from `1-dim` means copy from `1-3` if we are in a 3 dimensional gather?